protoc-gen-go/grpc/grpc.go

```diff
@@ -223,6 +223,7 @@ func (g *grpc) generateService(file *generator.FileDescriptor, service *pb.Servi
 	}
 
 	// Service descriptor.
+	g.P("func Get", servName, "ServiceDescriptor() ([]byte, []int) { return ", file.VarName(), ", []int{", strconv.Itoa(index), "} }")
```

Comment:
Observation: it's technically possible for this to collide with another symbol in the proto file that ends up defined as a top-level type.  This is also true for the `New` functions we generate.  Is that OK?

Note that if we use `<MyService>Descriptor()`, we're probably more likely to collide than prepending the "Get".  Another option: `<MyService>_Descriptor()`.

Reply:
I don't have an opinion here. Collision is possible but unlikely because I don't see people naming services FooServiceServiceDescriptor.

Reply:
I don't see collision as highly possible, either. It seems difficult to define a go proto package that defines two similarly named proto `service`s. And then I also think it would be difficult to make the go compiler happy. It's unfortunate and I think it would be a problem better resolved by the code author. I really don't expect there to be many situations where this collides.

Reply:
> I don't see people naming services FooServiceServiceDescriptor.

A message named `FooServiceDescriptor` doesn't seem too absurd.  But if we're going with `Get`, then it's much more unlikely.  And `NewFooService` also seems plausible.

@dsnet & @paranoiacblack, my question remains: can we not find a way for the main Go code generator to support more easily accessing ServiceDescriptorProtos?  It seems unusual to me for the grpc plugin to be the thing that provides this capability since it is a staadard proto feature.

Reply:
I'm not sure what you mean by supporting more easy access to service descriptor protos; this change seems to be the easiest access I've heard suggested so far. The other approach we discussed was exporting file descriptors, which would allow users to get to the nested service descriptors. While it might be nice to do that, I don't think it would be easier for users.

I don't think it's odd for the gRPC plugin to generate this code, currently, since it is the defacto proto service implementation for RPCs. If there were another popular implementation, we could move this logic to the main protoc-gen-go to generate generic service descriptors. It seems okay for the gRPC plugin to do this because there's no particular reason to generate this unless grpc is invoked, anyways.

Reply:
If popularity is measured by github stars, twirp is super popular: https://github.com/twitchtv/twirp. I wouldn't be surprised if it overtakes grpc by the end of the year.

Reply:
> this change seems to be the easiest access I've heard suggested so far

These same functions could just as easily be part of the main Go code generator.  Service descriptor protos are not grpc-specific; they're "services" specific.  I see this as a part of basic introspection of proto definitions.

If you consider this idea and still believe this belongs in only the grpc plugin, then let's just merge this and move forward.

Reply:
I don't think this only belongs in the gRPC plugin. On the contrary, I'm only stating that it is not unusual, as you originally claimed. This PR is particularly about being able to do this with gRPC. If y'all really foresee this being required/useful for twirp or some other RPC implementation, then let's go ahead and export this from the protoc plugin, directly, I suppose.

Reply:
I'd like to merge this PR if possible. @paranoiacblack I don't think there any remaining blockers.

Reply:
Created #579 for doing this in the main generator instead.

Note that this PR is into the master branch, which the proto team does not allow any longer.